### PR TITLE
Updates to supervised mcts evaluation

### PR DESCRIPTION
**Pull request by noahred16**
I noticed potentially a couple issues with the supervised-mcts evaluation. I figured since I'd be updating your code, I would submit a PR for you to review. Feel free to check out this branch and comment on any changes. Feedback is welcome. 

### Issues:
1.) Primarily, 2 of the 3 eval functions were not using the mcts `search` function but were instead using the `predict` function, which basically doesn't make use of mcts at all.  
2.) The test data you were using was not necessarily the same exactly 20% that was excluded from the data used to train the model.  
  
### Updates:  
- In previous commit, I separated the Connect4Net class and DataTransformer class out so that it could be reused. 
- Updated evaluate_supervised_mcts_on_test_data to use the search function. Also made sure to test on the same test set that was excluded during training. 
- Made use of seeds for reproducibility
- For evaluation, "correct" is true for any move that has the same value as the best move. 
- Added loaders for the evaluation functions, bc they take a while. 

### Results
I'm not sure we are all the way there yet for supervised-mcts. I'll need to look more at the rollout step to make sure the valuation is correct. Will want to look into speeding things up, because the eval methods take a while rn. It makes sense. For each board, the mcts algo has to simulate 800 episodes to produce a tree.

### CODE REVIEW

**Comment by noahred16** [comment]
Instead of using num_visits / total_visits to determine the output of mcts, I'm thinking we may want to use the average score instead (child.total_score / child.num_visits). This thought is because moves close to a win will be explored but have few visits because they are terminal leafs and no longer get considered.

**Comment by noahred16** [comment]
I'm not sure though. I havn't been able to see AlphaZero say anything like this. Most resources say to use the number of visits. But I don't understand how that makes sense. Why would one continuously visit a state if the value is already known.

**Comment by noahred16** [comment]
I wrote this test to explore the idea. I added a root prop to the mcts class so we could print out the children values for debugging. 

```
def test_sup_mcts_search():
    board = np.array(
        [
            [0, 0, 0, 0],
            [1, -1, 0, 0],
            [1, -1, 0, 0],
            [1, -1, 0, 0],
        ],
        dtype=int,
    )
    game = Connect4(num_of_rows=4, num_of_cols=4, board=board)
    model_path = "models/connect4_4x4_supervised.pt"
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    iterations = 500
    supervised_mcts = SupervisedMCTS(
        model_path=model_path, device=device, iterations=iterations
    )

    move_probs = supervised_mcts.search(game)

    print("Move probabilities:", move_probs)

    # # argmax
    best_move = np.argmax(move_probs)
    # print("Best move found by Supervised MCTS:", best_move)


    for move, child in supervised_mcts.root.children.items():
        print(f"Move: {move}, Visits: {child.num_visits}, Score: {child.total_score}")

    # assert best_move is not None
    assert best_move == 0
```

**Comment by lee-nux** [comment]
> Instead of using num_visits / total_visits to determine the output of mcts, I'm thinking we may want to use the average score instead (child.total_score / child.num_visits). This thought is because moves close to a win will be explored but have few visits because they are terminal leafs and no longer get considered.

I actually tried this so far, the accuracy seems to be much worse with a whopping 55% unfortunately, maybe tweaking the hyperparamters might help! I'm thinking about dropping the scaling the wins to favor quicker wins next to see if that's the problem

**Comment by lee-nux** [comment]
dropping scaling took us to about 87% accuracy, i did a little bit of research, it seems like having a clear +1 for win 0 for draw and -1 lose is the traditional way to go about it for these games, maybe we should stick to this! Let me know your thoughts!
